subprojects/model-core/src/main/java/org/gradle/internal/properties/annotations/AbstractTypeMetadataWalker.java

```diff
@@ -167,5 +193,13 @@ private static <T> TypeToken<?> extractNestedType(TypeToken<T> beanType, Class<?
             ParameterizedType type = (ParameterizedType) beanType.getSupertype(parameterizedSuperClass).getType();
             return TypeToken.of(type.getActualTypeArguments()[typeParameterIndex]);
         }
+
+        @Override
+        Set<TypeToken<?>> newIdentitySetOf(Set<TypeToken<?>> initialSet, TypeToken<?> newElement) {
+            return ImmutableSet.<TypeToken<?>>builder()
+                .addAll(initialSet)
+                .add(newElement)
+                .build();
+        }
```

Comment:
This seems quite inefficient. We could instead use a mutable `Map` and add/remove elements as we go.

Reply:
Now I use a single instance and do add/remove.